File: packages/nexus-coordinator/src/nexus_coordinator/invite.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""Coordinator-side invite minting, tracking, and revocation.

The wire format itself lives in the Rust crate
``nexus-worker-core::invite`` and is reached through
:func:`nexus_core.mint_invite` / :func:`nexus_core.decode_invite`.
This module adds the **persistence** layer the Rust crate does
not care about:

- ``invites`` SQLite table with (id, wire, scope, expires_at,
  max_uses, uses_count, revoked_at, note) so the coordinator can
  list / revoke / cap invites at the Python API boundary.
- :class:`InviteLedger` wraps CRUD + validity checks.

The Rust decoder does the cryptographic work (signature +
expiry + Worker-requires-ticket); this ledger layers the
"coordinator-issued and not revoked" gate on top.
"""

from __future__ import annotations

import time
import uuid
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import aiosqlite
import nexus_core
import structlog

_log = structlog.get_logger(__name__)

_INVITES_SCHEMA = """
CREATE TABLE IF NOT EXISTS invites (
    id            TEXT PRIMARY KEY,
    wire          TEXT NOT NULL UNIQUE,
    scope         TEXT NOT NULL,
    project_id    TEXT NOT NULL,
    project_name  TEXT NOT NULL,
    expires_at    INTEGER NOT NULL,
    max_uses      INTEGER,
    uses_count    INTEGER NOT NULL DEFAULT 0,
    revoked_at    INTEGER,
    note          TEXT,
    created_at    INTEGER NOT NULL
);

CREATE INDEX IF NOT EXISTS invites_by_revoked ON invites(revoked_at);
CREATE INDEX IF NOT EXISTS invites_by_expires_at ON invites(expires_at);
"""


@dataclass
class InviteRecord:
    """A single row from the ``invites`` table."""

    id: str
    wire: str
    scope: str
    project_id: str
    project_name: str
    expires_at: int
    max_uses: int | None
    uses_count: int
    revoked_at: int | None
    note: str | None
    created_at: int


class InviteLedger:
    """CRUD + validity tracking for coordinator-issued invites."""

    def __init__(self, *, db_path: Path, coord_secret: bytes) -> None:
        self._db_path = db_path
        self._coord_secret = coord_secret

    async def init(self) -> None:
        async with aiosqlite.connect(self._db_path) as db:
            await db.executescript(_INVITES_SCHEMA)
            await db.commit()

    async def mint(
        self,
        *,
        project_id: str,
        project_name: str,
        scope: str,
        tasks_doc_ticket: str | None,
        expiry_secs: int,
        coordinator_addr: str | None = None,
        max_uses: int | None = None,
        note: str | None = None,
    ) -> InviteRecord:
        """Mint a new invite and persist it. Returns the full record.

        Validates scope + ticket combination before calling the
        Rust mint function, so a caller error (worker scope
        without ticket) is surfaced as a ``ValueError`` with the
        same message shape as the Rust side.
        """
        if scope not in ("worker", "observer"):
            raise ValueError(f"scope must be 'worker' or 'observer', got {scope!r}")
        if scope == "worker" and not tasks_doc_ticket:
            raise ValueError("tasks_doc_ticket is required when scope == 'worker'")
        if expiry_secs <= 0:
            raise ValueError(f"expiry_secs must be positive, got {expiry_secs}")

        now = int(time.time())
        expires_at = now + expiry_secs
        wire = nexus_core.mint_invite(
            self._coord_secret,
            project_id,
            project_name,
            coordinator_addr,
            tasks_doc_ticket,
            scope,
            expires_at,
        )

        record = InviteRecord(
            id=f"inv-{uuid.uuid4().hex[:16]}",
            wire=wire,
            scope=scope,
            project_id=project_id,
            project_name=project_name,
            expires_at=expires_at,
            max_uses=max_uses,
            uses_count=0,
            revoked_at=None,
            note=note,
            created_at=now,
        )

        async with aiosqlite.connect(self._db_path) as db:
            await db.execute(
                """
                INSERT INTO invites (
                    id, wire, scope, project_id, project_name, expires_at,
                    max_uses, uses_count, revoked_at, note, created_at
                ) VALUES (?, ?, ?, ?, ?, ?, ?, 0, NULL, ?, ?)
                """,
                (
                    record.id,
                    record.wire,
                    record.scope,
                    record.project_id,
                    record.project_name,
                    record.expires_at,
                    record.max_uses,
                    record.note,
                    record.created_at,
                ),
            )
            await db.commit()

        _log.info(
            "invite minted",
            invite_id=record.id,
            scope=scope,
            project_id=project_id,
            expires_at=expires_at,
        )
        return record

    async def revoke(self, invite_id: str) -> bool:
        """Mark an invite as revoked. Returns True if the row
        existed and was not already revoked."""
        now = int(time.time())
        async with aiosqlite.connect(self._db_path) as db:
            cursor = await db.execute(
                "UPDATE invites SET revoked_at = ? WHERE id = ? AND revoked_at IS NULL",
                (now, invite_id),
            )
            affected = cursor.rowcount
            await cursor.close()
            await db.commit()
        return bool(affected)

    async def list_invites(self, limit: int = 100) -> list[InviteRecord]:
        async with aiosqlite.connect(self._db_path) as db:
            async with db.execute(
                """
                SELECT id, wire, scope, project_id, project_name, expires_at,
                       max_uses, uses_count, revoked_at, note, created_at
                FROM invites ORDER BY created_at DESC LIMIT ?
                """,
                (limit,),
            ) as cursor:
                rows = await cursor.fetchall()
        return [
            InviteRecord(
                id=row[0],
                wire=row[1],
                scope=row[2],
                project_id=row[3],
                project_name=row[4],
                expires_at=row[5],
                max_uses=row[6],
                uses_count=row[7],
                revoked_at=row[8],
                note=row[9],
                created_at=row[10],
            )
            for row in rows
        ]

    async def get(self, invite_id: str) -> InviteRecord | None:
        rows = await self.list_invites(limit=10_000)
        return next((r for r in rows if r.id == invite_id), None)

    @staticmethod
    def decode(wire: str, *, now_unix: int | None = None) -> dict[str, Any]:
        """Parse an ``nx1...`` string, verify signature + expiry,
        and return the payload fields as a plain dict."""
        return nexus_core.decode_invite(wire, now_unix)

File: crates/nexus-coordinator-rs/src/invite.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
//! Invite minting, tracking, and revocation ledger
//! (Sprint 41 Phase B, port of invite.py S22).

use std::time::{SystemTime, UNIX_EPOCH};

use serde::{Deserialize, Serialize};

use crate::db::CoordinatorDb;
use crate::error::CoordinatorError;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InviteRecord {
    pub id: String,
    pub wire: String,
    pub scope: String,
    pub project_id: String,
    pub project_name: String,
    pub expires_at: i64,
    pub max_uses: Option<i64>,
    pub uses_count: i64,
    pub revoked_at: Option<i64>,
    pub note: Option<String>,
    pub created_at: i64,
}

pub struct MintRequest<'b> {
    pub id: &'b str,
    pub wire: &'b str,
    pub scope: &'b str,
    pub project_id: &'b str,
    pub project_name: &'b str,
    pub expires_at: i64,
    pub max_uses: Option<i64>,
    pub note: Option<&'b str>,
}

pub struct InviteLedger<'a> {
    db: &'a CoordinatorDb,
}

impl<'a> InviteLedger<'a> {
    pub fn new(db: &'a CoordinatorDb) -> Self {
        Self { db }
    }

    pub fn mint(&self, req: &MintRequest<'_>) -> Result<InviteRecord, CoordinatorError> {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs() as i64;
        self.db.conn().execute(
            "INSERT INTO invites \
             (id, wire, scope, project_id, project_name, expires_at, \
              max_uses, uses_count, revoked_at, note, created_at) \
             VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, 0, NULL, ?8, ?9)",
            rusqlite::params![
                req.id,
                req.wire,
                req.scope,
                req.project_id,
                req.project_name,
                req.expires_at,
                req.max_uses,
                req.note,
                now,
            ],
        )?;
        Ok(InviteRecord {
            id: req.id.to_string(),
            wire: req.wire.to_string(),
            scope: req.scope.to_string(),
            project_id: req.project_id.to_string(),
            project_name: req.project_name.to_string(),
            expires_at: req.expires_at,
            max_uses: req.max_uses,
            uses_count: 0,
            revoked_at: None,
            note: req.note.map(String::from),
            created_at: now,
        })
    }

    pub fn revoke(&self, invite_id: &str) -> Result<bool, CoordinatorError> {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs() as i64;
        let changed = self.db.conn().execute(
            "UPDATE invites SET revoked_at = ?1 WHERE id = ?2 AND revoked_at IS NULL",
            rusqlite::params![now, invite_id],
        )?;
        Ok(changed > 0)
    }

    pub fn get(&self, invite_id: &str) -> Result<Option<InviteRecord>, CoordinatorError> {
        let result = self
            .db
            .conn()
            .query_row(
                "SELECT id, wire, scope, project_id, project_name, expires_at, \
                 max_uses, uses_count, revoked_at, note, created_at \
                 FROM invites WHERE id = ?1",
                rusqlite::params![invite_id],
                row_to_invite,
            )
            .ok();
        Ok(result)
    }

    pub fn list(&self, limit: usize) -> Result<Vec<InviteRecord>, CoordinatorError> {
        let conn = self.db.conn();
        let mut stmt = conn.prepare(
            "SELECT id, wire, scope, project_id, project_name, expires_at, \
             max_uses, uses_count, revoked_at, note, created_at \
             FROM invites ORDER BY created_at DESC LIMIT ?1",
        )?;
        let rows = stmt.query_map(rusqlite::params![limit as i64], row_to_invite)?;
        let mut result = Vec::new();
        for row in rows {
            result.push(row?);
        }
        Ok(result)
    }
}

fn row_to_invite(row: &rusqlite::Row<'_>) -> rusqlite::Result<InviteRecord> {
    Ok(InviteRecord {
        id: row.get(0)?,
        wire: row.get(1)?,
        scope: row.get(2)?,
        project_id: row.get(3)?,
        project_name: row.get(4)?,
        expires_at: row.get(5)?,
        max_uses: row.get(6)?,
        uses_count: row.get(7)?,
        revoked_at: row.get(8)?,
        note: row.get(9)?,
        created_at: row.get(10)?,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_db() -> CoordinatorDb {
        CoordinatorDb::open_in_memory().unwrap()
    }

    fn mk_req<'b>(id: &'b str, wire: &'b str, scope: &'b str) -> MintRequest<'b> {
        MintRequest {
            id,
            wire,
            scope,
            project_id: "proj1",
            project_name: "TestProject",
            expires_at: 9999999999,
            max_uses: None,
            note: None,
        }
    }

    #[test]
    fn mint_and_get() {
        let db = make_db();
        let ledger = InviteLedger::new(&db);
        let mut req = mk_req("inv-001", "nx1abc", "worker");
        req.max_uses = Some(10);
        req.note = Some("test invite");
        let rec = ledger.mint(&req).unwrap();
        assert_eq!(rec.id, "inv-001");
        assert_eq!(rec.scope, "worker");
        assert_eq!(rec.uses_count, 0);
        assert!(rec.revoked_at.is_none());
        let fetched = ledger.get("inv-001").unwrap().unwrap();
        assert_eq!(fetched.wire, "nx1abc");
    }

    #[test]
    fn revoke_invite() {
        let db = make_db();
        let ledger = InviteLedger::new(&db);
        ledger.mint(&mk_req("inv-r1", "wire1", "observer")).unwrap();
        assert!(ledger.revoke("inv-r1").unwrap());
        let rec = ledger.get("inv-r1").unwrap().unwrap();
        assert!(rec.revoked_at.is_some());
        assert!(!ledger.revoke("inv-r1").unwrap());
    }

    #[test]
    fn list_invites() {
        let db = make_db();
        let ledger = InviteLedger::new(&db);
        ledger.mint(&mk_req("inv-a", "w1", "worker")).unwrap();
        ledger.mint(&mk_req("inv-b", "w2", "observer")).unwrap();
        let list = ledger.list(10).unwrap();
        assert_eq!(list.len(), 2);
    }

    #[test]
    fn get_nonexistent() {
        let db = make_db();
        let ledger = InviteLedger::new(&db);
        assert!(ledger.get("nope").unwrap().is_none());
    }
}
